Add support for iterative pruning based on user-specified discard conditions

ns-rse opened the pull request:
First stab at iterative pruning.

ns-rse committed:
Adds functionality to iteratively prune paths

Introduces the function `iteratively_prune_paths()` which will prune paths of type 1 and 3 until there are a user
specified (via `min_skeleton`) number of paths left.

* Documentation on pruning skeleton included.
* Retains the `keep_images` option by adding it as an attribute if supplied. Required because the `.prune_paths()`
  method when applied returns the pruned skeleton but didn't retain this option (as far as I could work out).
* Tests on dummy skeletons are included

ns-rse committed:
Pruning loops and linear skeletons to retain longest path

Refines the steps taken in pruning such that large loops are retained as the target if present or if skeletons are
linear then the longest path is retained.

Tests are included for a range of dummy skeletons.

ns-rse committed:
Loops and linear skeletons retain longest path

Refines the steps taken in pruning such that large loops are retained as the target if present or if skeletons are
linear then the longest path is retained.

Tests are included for a range of dummy skeletons.

ns-rse committed:
Adding examples for pruning

ns-rse committed:
Adding examples for pruning

ns-rse committed:
Merge branch 'main' into ns-rse/iterative-pruning

ns-rse committed:
Tweak to text

jni committed:
Add function to make nx Graph from Skeleton

jni committed:
Add function to compute main branches in nx directly

jni committed:
Initial work on networkx-based pruning

jni commented:
I've made some changes to the code, still needs work but hopefully enough to get the gist.

Ultimately, I think we need to be able to go from Skeleton to Graph and back losslessly. Like sparse array formats, each representation has strengths and weaknesses and by "allowing" either format we enable different algorithms.

ns-rse committed:
Networkx <--> Skeleton development and tests

I was looking to convert NetworkX objects back to Skeletons and stumbled on the `nx.to_numpy_array()` function and the
related `nx.from_numpy_array()`.

Since `csr.skeleton_image` are typically Numpy Arrays I figured this would be a convenient and efficient method of
converting between the two formats so have attempted to replace `skeleton_to_nx()` with this.

I had written some tests for `skeleton_to_nx()` first before making undertaking refactoring.

I found that `nx.from_numpy_array()` doesn't handle non-square arrays, ok not a problem we can pad them with
zeros (wrote a small function to do this along with tests), but on applying `nx.from_numpy_array()` to these padded
arrays the tests I wrote to originally failed, which surprised me as, in my naivety, I wouldn't expect padding with
zero's to adjacency graphs to affect the resulting network representation in terms of the number of nodes and the number
of edges.

jni commented:
@ns-rse looking at the description and implementation (not tests) of 0ad90df, I don't see how it could ever work, because a skeleton image is *not* what `to_numpy_array` is producing: it produces an adjacency matrix — and loses all our metadata (such as the path between junctions and tips) in the process.

The naive way to produce the image is to make an empty image, then iterate over all the edges and set the edge coordinates to 1 in the image, before passing that to `Skeleton`. I have vague hopes of finding something faster by going directly from the edges to the csr_matrix for paths, but that sounds complicated — and can be left to a later optimisation.

ns-rse commented:
Yes, I'm realising I have deficiency in my udnerstanding of graph theory NetworkX and how Skan should be interacting with it. The tests are hopefully going to be useful and your code to go from `Skeleton` to `NetworkX` graph work but I need to rethink how to get back from NetworkX to Skeletons whilst preserving the metadata (and also be mindful that we might prune the graph under NetworkX). Hope to spend some more time on it later this week.

ns-rse committed:
Adds array_to_nx() and nx_to_array() functions

Introduces conversion of Numpy arrays (_not_ adjacency matrices) to NetworkX graphs.

This is done by assuming that each element of a 2-D boolean array[1] is a node connected by edges, we look at what
values the neighbours are and > 0 the two points form an edge which is added to a NetworkX graph.

The shape of the original Numpy array is saved within the `Networkx.graph` object which makes
reconstructing the original Numpy array straight-forward (as we may be pruning some of the original co-ordinates may be
removed).

Thanks to @MaxGamill-Sheffield for guiding me to this approach and as noted a solution found on SO for how to find the
co-ordinates of neighbouring cells.

ns-rse committed:
Correcting doc string

ns-rse committed:
Merge branch 'main' into ns-rse/iterative-pruning

ns-rse committed:
Correcting import order

ns-rse committed:
Correcting test, had chopped part of key fixture resolving conflicts

jni commented:
> The solution in https://github.com/jni/skan/commit/fbc09886a3c70f5c174f23d2de3c209a95938390 treats every element of a 2-D array as a node and makes going back and forth between the two straight-forward.

Ah, sorry, I didn't have time to review this at the time.

I don't think that's going to fly unfortunately. Actually, a bit of lore: all the main implementations of skan are in a file called `csr.py` because the very earliest implementations were in networkx. But we found that was suuuuper slow so I went about methodically converting everything to `scipy.sparse.csr_matrix` and writing out skan as it is today. But while I was working on that it was siloed in this little `csr.py` file, until that file became the whole thing. 😊 

So, I think that the 2D pixel graph is too low level to try to do a round trip — we will be swapping one performance issue for another. The correct level of abstraction for the Skeleton, imho, is the junction-junction graph, which is what you get when you make use the node id pairs (src, dst) from the `summarize` table as an edge list. Then it's a matter of putting the right metadata on the networkx graph. For example, each row of the `summarize` table has a row index, which corresponds to `path_coordinates(idx)` on the `Skeleton` object — so we can store the coordinates (or maybe just the pixel IDs) as an edge attribute. Then, when you prune, you need to merge two edges, and the path coordinates of the new edge is just `np.concatenate([edge1['coords'], edge2['coords'][1:]], axis=0)`, for example. All of this gives you enough information to reconstruct a new Skeleton object, in the worst case by writing out the pixels and recomputing a csr skeleton graph — but maybe faster, I need to think about it.

ns-rse committed:
Rebase on main

Adapting to NetworkX graphs.

Currently tests fail for src/skan/test/test_csr.py::test_iteratively_prune_[paths.multiple_paths] as there is no
'discard' argument. Will revisit and try and work this out, been so long I don't understand how this argument evaluates
to boolean and how to construct the tests.

ns-rse commented:
I've been having a play around with this on [ns-rse/networkx](https://github.com/ns-rse/skan/tree/ns-rse/networkx) and had a go at passing the coordinates as edge attributes to NetworkX and reonconstructing paths between co-ordinates using Bresnham's Line algorithm.

It works in some instances but in writing tests I discovered there is insufficient data retained. For example `skeleton1` in `src/skan/_testdata.py` has a loop in it. This results in two nodes with the same co-ordinates but there is no information about what the points are connecting them.

```python
from skan import Skeleton, summarize
from skan._testdata import skeleton1

summarize(Skeleton(skeleton1))[['skeleton-id', 'coord-src-0', 'coord-src-1', 'coord-dst-0', 'coord-dst-1']]

   skeleton-id  coord-src-0  coord-src-1  coord-dst-0  coord-dst-1
0            0            2            1            3            3
1            0            2            1            3            3
2            0            2            1            4            0
3            0            3            3            4            6
```

...and so we've no idea how those two points differ in terms of the path taken between them. This is going to be true of any non-linear edge and so whilst we could recombine the path co-ordinates of edges after pruning in Networkx as you suggested we would have no idea the path taken between those points.

The only solution I can think of so far that doesn't involve using every point as a node is to pass the co-ordinates of the points that make up an edge as an attribute to NetworkX and concatenate those after pruning. Not tried this yet and have been asked to work on something else in the coming weeks but will think about it and try and find some time to try this out.

jni commented:
@ns-rse sorry about the long wait here, as it has been very busy trying to get napari 0.4.19 out the door (still not done) and also family stuff. But I haven't forgotten this!

Quick thoughts:

> reonconstructing paths between co-ordinates using Bresnham's Line algorithm.

I don't think this should be needed because I think we should keep the path coordinates and they should be pixel-perfect — there should be no need to infer any pixels. You need to keep the info from *both* the summary *and* the skeleton using `path_coordinates(i)` where i is the path ID of each row in the summary. iirc I got those matched up exactly at some point. 

> The only solution I can think of so far that doesn't involve using every point as a node is to pass the co-ordinates of the points that make up an edge as an attribute to NetworkX and concatenate those after pruning. Not tried this yet and have been asked to work on something else in the coming weeks but will think about it and try and find some time to try this out.

Yes like that. But that still vastly reduces the number of nodes in the graph.

jni commented:
I will try to come back to this in the next couple of weeks but didn't want to leave you hanging any longer, again, sorry about the wait!

ns-rse commented:
Don't worry at all @jni we all have lives. Hope you managed some time away from the computer.

That is however super-useful I hadn't clocked `path_coordinates()` method and think that would be the perfect tool to for adding the co-ordinates as an attribute.

I'll have a tinker next week and see how I get on.

Thank you and Happy New Year

jni commented:
@ns-rse would you mind rebasing this on main? I can do it if you are too busy right now though.

ns-rse committed:
Merge branch 'main' into ns-rse/iterative-pruning

ns-rse commented:
@jni I merged main in and solved the conflicts, tests failed and so I tried to sort those out locally. Fixed the failing ones but there were more that failed and I haven't got them all sorted yet though as I don't understand how the `discard` option evaluates to `boolean` for the `itteratively_prune_paths()` function. Will try and take a look over the coming days and work that out.

jni committed:
Add full data dict to graph attributes

jni committed:
Use multi-graph edges with keys everywhere

Even when there is only one edge, for consistency, every edge in a multi-graph also contains a key (starting at 0). We use the key everywhere, which makes the code a bit more complicated if you don't expect multigraph conventions, but more consistent overall.

jni committed:
Add example predicate with is_endpoint

jni committed:
Handle self-edges / redundant edges

jni committed:
Updates to iterative pruning example

jni commented:
I'm almost there, but I think my current approach is too gung-ho... 😂 

<img width="734" alt="Screenshot 2024-06-29 at 11 57 46 AM" src="https://github.com/jni/skan/assets/492549/64761851-847d-4b87-b1ed-66e08beba41b">

jni committed:
Merge branch 'main' into ns-rse/iterative-pruning

jni commented:
Ah, the hardest problem in computer science — off by one errors. This is one of the intermediate steps:

<img width="1198" alt="Screenshot 2024-06-29 at 5 19 26 PM" src="https://github.com/jni/skan/assets/492549/bce56943-6314-44c5-8227-c4cc82d972db">

<img width="1198" alt="Screenshot 2024-06-29 at 5 18 30 PM" src="https://github.com/jni/skan/assets/492549/897f17fe-48a7-4bdb-bc82-cc7fc6f830ab">

So when removing some of the branches we are breaking up the inner loop, at which point we're already doomed. 😂 Otherwise it's looking pretty good though. Just need to figure out how to avoid cutting open the middle branches...

jni commented:
btw @ns-rse it's a cool data sample — can you tell me what the xy pixel spacing/resolution is? We can incorporate that into the example.

jni committed:
Use minimum buffer size of 1 to support empty ims

jni committed:
Use keep_images=True when creating skeletons

This is useful for debugging - can make a parameter later

jni committed:
Include values an indices in merged edges

jni committed:
Add todo note for remaining attributes

ns-rse commented:
Hi @jni 

> Ah, the hardest problem in computer science — off by one errors. This is one of the intermediate steps:

:laughing: 

I encountered this with the linear molecules though, it just kept on pruning everything back. I think I solved it by looking for the longest path, but not sure how that transfers to loops/closed structures where a break is then introduced.

> btw @ns-rse it's a cool data sample — can you tell me what the xy pixel spacing/resolution is? We can incorporate that into the example.

The pixel to nanometre scaling on these images varies depending on the resolution used in the scanning. I think this image is from [`minicircle.spm`]() where the scaling is `0.4940029296875` 

Thanks for continuing to hack away at this :+1: 

Let me know if there is anything I can help with.

jni commented:
> 0.4940029296875

... nm/pixel? ... That's cool/crazy! 😃

jni committed:
Minor changes

jni committed:
Merge branch 'main' into ns-rse/iterative-pruning

jni commented:
Ok I figured out the bug (or at least *a* bug). In this code:

```python
    middle_node = (set(e1[:2]) & set(e2[:2])).pop()
    new_edge = sorted(
            (set(e1[:2]) | set(e2[:2])) - {middle_node},
            key=lambda i: i in e2[:2],
            )
    if len(new_edge) == 1:  # self-edge
        new_edge = (new_edge[0], new_edge[0])
    d1 = g.edges[e1]
    d2 = g.edges[e2]
    p1 = d1['path'] if e1[1] == middle_node else d1['path'][::-1]
    p2 = d2['path'] if e2[0] == middle_node else d2['path'][::-1]
    i1 = d1['indices'] if e1[1] == middle_node else d1['indices'][::-1]
    i2 = d2['indices'] if e2[0] == middle_node else d2['indices'][::-1]
    v1 = d1['values'] if e1[1] == middle_node else d1['values'][::-1]
    v2 = d2['values'] if e2[0] == middle_node else d2['values'][::-1]
```

we assume that the edges (i, j) given by the iterator in `nx.Graph.edges()` are given in the order in which they are added, that is, if we added (j, i) it should reappear as (j, i) and not (i, j). However, that's not necessarily the case, because the graph is not directed. Therefore, the path coordinates can appear in any order, and sorting them based on the input edge IDs is not generally correct.

The fix is to check the ordering of the paths based on the indices rather than the edge IDs. Will work on that now. 🤞 

I don't think this explains why the full graph gets removed, because the logic is entirely within the networkx graph. However, it explains why connection points were being removed. I think that where the nx logic is probably going wrong is if you end up with (i, j) and (j, i) as the two edges — this is a multigraph, so it's allowed), and then it gets treated not as a loop but as two paths between the same nodes — and the less bright one gets removed. Will check that next.

jni commented:
> I think that where the nx logic is probably going wrong is if you end up with (i, j) and (j, i) as the two edges — this is a multigraph, so it's allowed), and then it gets treated not as a loop but as two paths between the same nodes

Hmmm, no, `_remove_simple_path_nodes` accounts for this. But maybe it does so incorrectly. Will see. One fix at a time. 😅

jni committed:
Check path matches based on indices, not nx IDs

jni commented:
Ok better! But the whole thing still disappears. I think it's during the removing-simple-nodes loop. But, getting closer!

![CleanShot 2025-05-12 at 18 01 03@2x](https://github.com/user-attachments/assets/4ec42a41-5796-45b5-9435-bf71c127abc2)

ns-rse commented:
Thanks for continuing to work on this @jni.

It's been a while since I've thought about this in any great detail.

I think you're right that the predicates may get things wrong when there are loops and increasingly imaging of interacting molecules has been a major area of interest (its what instigated the work on pruning in the first place).

Keeping things simple for users is also a big bonus and so whilst it might be slower I'd be inclined to go with the sequential removal of edges.

I've let other members of the research group know about this though and maybe they might have some thoughts. @maxgamill-sheffield in particular might have useful insight/thoughts.

jni committed:
Remove one edge at a time

jni committed:
Update docs to new functions

jni committed:
Update napari pyqt testing requirement and add zarr

jni committed:
Add new testdata

jni committed:
Iteratively prune paths with a float priority function

jni committed:
Rename spur length to be simpler

jni committed:
Remove unused code

jni committed:
Merge branch 'main' into ns-rse/iterative-pruning

jni committed:
Remove failing test as a Skeleton with one pixel is no longer a ValueError
